src/Bicep.Core/TypeSystem/DeployTimeConstantVariableVisitor.cs

```diff
@@ -0,0 +1,122 @@
+        }
+
+        public override void VisitVariableDeclarationSyntax(VariableDeclarationSyntax syntax)
+        {
+            if (this.invalidReferencedBodyObj != null)
+            {
+                return;
+            }
+            visitedStack.Push(syntax.Name.IdentifierName);
+            base.Visit(syntax.Value);
+            if (this.invalidReferencedBodyObj == null)
+            {
+                visitedStack.Pop();
+            }
+        }
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            if (this.invalidReferencedBodyObj != null)
+            {
+                return;
+            }
+            base.VisitObjectPropertySyntax(syntax);
+        }
+
+        public override void VisitArrayAccessSyntax(ArrayAccessSyntax syntax)
+        {
+            if (syntax.BaseExpression is VariableAccessSyntax baseSyntax)
+            {
+                var baseSymbol = model.GetSymbolInfo(baseSyntax);
+                switch(baseSymbol)
+                {
+                    // do not visit the baseExpression if we have a symbol match. Return instead of break
+                    case ResourceSymbol:
+                    case ModuleSymbol:
+                        if (TypeAssignmentVisitor.UnwrapType(((DeclaredSymbol)baseSymbol).Type) is ObjectType referencedBodyObj)
+                        {
+                            switch (syntax.IndexExpression)
+                            {
+                                case StringSyntax stringSyntax:
+                                if (stringSyntax.TryGetLiteralValue() is string property &&
+                                    referencedBodyObj.Properties.TryGetValue(property, out var propertyType) &&
+                                    !propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant))
+                                    {
+                                        this.invalidReferencedBodyObj = referencedBodyObj;
+                                        visitedStack.Push(baseSymbol.Name);
+                                    }
+                                    break;
+                            }
+                        }
+                        return;
+                }
+            }
+            base.Visit(syntax.BaseExpression);
+        }
+
+        public override void VisitPropertyAccessSyntax(PropertyAccessSyntax syntax)
+        {
+            if (syntax.BaseExpression is VariableAccessSyntax baseSyntax)
+            {
+                var baseSymbol = model.GetSymbolInfo(baseSyntax);
+                switch(baseSymbol)
+                {
+                    // do not visit the baseExpression if we have a symbol match. Return instead of break
+                    case ResourceSymbol:
+                    case ModuleSymbol:
+                        if (TypeAssignmentVisitor.UnwrapType(((DeclaredSymbol)baseSymbol).Type) is ObjectType referencedBodyObj)
```

Comment:
It might be worth extracting the condition as a static helper function since it's repeated.

Reply:
Good catch, this is done 5 times.